modules/tensor_mechanics/src/materials/Compute2DFiniteStrain.C

```diff
@@ -18,12 +18,40 @@ validParams<Compute2DFiniteStrain>()
   InputParameters params = validParams<ComputeFiniteStrain>();
   params.addClassDescription(
       "Compute a strain increment and rotation increment for finite strains in 2D geometries.");
+
+  MooseEnum outOfPlaneDirection("x y z", "z");
+  params.addParam<MooseEnum>(
+      "out_of_plane_direction", outOfPlaneDirection, "The direction of the out-of-plane strain.");
   return params;
 }
 
 Compute2DFiniteStrain::Compute2DFiniteStrain(const InputParameters & parameters)
-  : ComputeFiniteStrain(parameters)
+  : ComputeFiniteStrain(parameters),
+    _out_of_plane_direction(getParam<MooseEnum>("out_of_plane_direction"))
+{
+}
+
+void
+Compute2DFiniteStrain::initialSetup()
 {
+  for (unsigned int i = 0; i < 3; ++i)
```

Comment:
3 -> LIBMESH_DIM

Reply:
I can try this but I don't know if it will work in these cases if this approach will work.  To get anything to work with my meshes when they are not in the x-y plane they must be output from cubit as 3d to be properly visualized in Paraview for example.  

Then running through MOOSE I require to have output_dimension = 3 in exodus block or the output exodus is a single line in the x-y plane that is output.  Therefore, it appears that once read in to MOOSE it is read as a 2D mesh so LIBMESH_DIM=2 and I have to loop over 3 not 2 to cover all the potential options.

Reply:
There is also something I saw related to the displacements vector in the [Mesh] block that caused me some troubles when originally with this PR.  Even though these are 2D problems as they are in a single plane to get proper answers in Finite strain (run on the displaced mesh) supplying displacements = 'disp_x disp_z' for example for analysis in the x-z plane on the displaced mesh these would be applied in x and y.  Therefore disp_y must be included as an AuxVariable with no AuxKernel and three displacements provided to the Mesh block for a 2D problem.